core/src/impl/Kokkos_Default_GraphNode_Impl.hpp

```diff
@@ -115,8 +127,17 @@ struct GraphNodeBackendSpecificDetails {
       // supported semantics, but instinct I have feels like it should be...
       m_has_executed = true;
       for (auto const& predecessor : m_predecessors) {
-        predecessor->execute_node();
+        predecessor->execute_node(exec);
       }
+
+      // Before executing the kernel, be sure to fence the execution space
+      // instance of predecessors.
+      for (const auto& predecessor : m_predecessors) {
+        if (predecessor->awaitable())
+          predecessor->get_execution_space().fence(
+              "Kokkos::DefaultGraphNode::execute_node: sync with predecessors");
+      }
```

Comment:
This is only necessary if the execution space instance the node was created with doesn't match the execution space instance used here, right?

Reply:
This won't be sufficient for thread-safety.

Reply:
A node cannot execute before all its predecessors are done.

The defaulted implementation can be thought of as if each node stores an execution space instance, potentially different from the one passed to `graph.submit(...)`.

So, before a node can actually launch its kernel, it has to fence its predecessors' execution space instances to ensure they are done.

We could save a fence or two with:
```suggestion
      // Before executing the kernel, be sure to fence the execution space
      // instance of predecessors.
      for (const auto& predecessor : m_predecessors) {
        if (predecessor->awaitable() && predecessor->get_execution_space() != this->get_execution_space())
          predecessor->get_execution_space().fence(
              "Kokkos::DefaultGraphNode::execute_node: sync with predecessors");
      }
```
if you want.

Reply:
Yes, I think that makes sense, in particular since you have already done something similar for https://github.com/kokkos/kokkos/pull/7271/files#r1742704039. I would assume the common case is for all nodes to execute on the same execution space instance and so there wouldn't be any fences.

Reply:
Done :heavy_check_mark: